File: zeus-backend/app/schemas/recipe.py
from pydantic import BaseModel, Field, validator
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class DifficultyLevel(str, Enum):
    EASY = "Easy"
    MEDIUM = "Medium"
    HARD = "Hard"


class MealType(str, Enum):
    BREAKFAST = "Breakfast"
    LUNCH = "Lunch"
    DINNER = "Dinner"
    SNACK = "Snack"
    DESSERT = "Dessert"


class Ingredient(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    quantity: str = Field("", max_length=100)  # Allow empty (e.g., "to taste" items)
    unit: str = Field("", max_length=50)  # Allow empty unit (e.g., "3 eggs")
    section: Optional[str] = Field(None, max_length=100)  # Recipe sub-section (e.g., "Salsa", "Tacos")


class Instruction(BaseModel):
    step: int = Field(..., ge=1)
    instruction: str = Field(..., min_length=1, max_length=2000)


class RecipeCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=300)
    description: Optional[str] = Field(None, max_length=2000)
    ingredients: List[Ingredient] = Field(..., min_items=1)
    instructions: List[Instruction] = Field(..., min_items=1)
    servings: int = Field(4, ge=1, le=20)
    prep_time: Optional[int] = Field(None, ge=0, le=480)  # max 8 hours
    cook_time: Optional[int] = Field(None, ge=0, le=480)  # max 8 hours
    cuisine_type: Optional[str] = Field(None, max_length=50)
    difficulty: DifficultyLevel = DifficultyLevel.MEDIUM
    meal_type: List[MealType] = Field(default=[])
    dietary_tags: List[str] = Field(default=[])
    image_url: Optional[str] = None

    @validator('instructions')
    def validate_instructions_order(cls, v):
        steps = [instruction.step for instruction in v]
        if steps != list(range(1, len(steps) + 1)):
            raise ValueError('Instructions must be numbered consecutively starting from 1')
        return v


class RecipeUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    ingredients: Optional[List[Ingredient]] = None
    instructions: Optional[List[Instruction]] = None
    servings: Optional[int] = Field(None, ge=1, le=20)
    prep_time: Optional[int] = Field(None, ge=0, le=480)
    cook_time: Optional[int] = Field(None, ge=0, le=480)
    cuisine_type: Optional[str] = Field(None, max_length=50)
    difficulty: Optional[DifficultyLevel] = None
    meal_type: Optional[List[MealType]] = None
    dietary_tags: Optional[List[str]] = None
    image_url: Optional[str] = None

    @validator('instructions')
    def validate_instructions_order(cls, v):
        if v is not None:
            steps = [instruction.step for instruction in v]
            if steps != list(range(1, len(steps) + 1)):
                raise ValueError('Instructions must be numbered consecutively starting from 1')
        return v


class RecipeResponse(BaseModel):
    id: str
    user_id: str
    title: str
    description: Optional[str]
    image_url: Optional[str]
    ingredients: List[Ingredient]
    instructions: List[Instruction]
    servings: int
    prep_time: Optional[int]
    cook_time: Optional[int]
    cuisine_type: Optional[str]
    difficulty: str
    meal_type: List[str]
    dietary_tags: List[str]
    is_ai_generated: bool
    likes_count: int
    created_at: datetime

    # Nutrition fields
    calories: Optional[int] = None
    protein_grams: Optional[float] = None
    carbs_grams: Optional[float] = None
    fat_grams: Optional[float] = None
    serving_size: Optional[str] = None

    # Optional fields for feed/detailed views
    creator_username: Optional[str] = None
    is_liked: Optional[bool] = None
    is_saved: Optional[bool] = None

    class Config:
        from_attributes = True


class RecipeFeedFilter(BaseModel):
    cuisine_type: Optional[str] = None
    cuisine_preferences: Optional[List[str]] = None
    difficulty: Optional[DifficultyLevel] = None
    max_difficulty: Optional[DifficultyLevel] = None
    max_prep_time: Optional[int] = Field(None, ge=0, le=480)
    meal_type: Optional[MealType] = None
    dietary_tags: Optional[List[str]] = None
    search: Optional[str] = None
    use_pantry_items: bool = False
    limit: int = Field(20, ge=1, le=500)
    offset: int = Field(0, ge=0)


class RecipeInteraction(BaseModel):
    recipe_id: str


class AIRecipeRequest(BaseModel):
    pantry_items: Optional[List[str]] = Field(default=[])
    cuisine_preference: Optional[str] = None
    dietary_restrictions: Optional[List[str]] = Field(default=[])
    cooking_skill: Optional[str] = Field(None, pattern="^(beginner|intermediate|advanced)$")
    max_prep_time: Optional[int] = Field(None, ge=5, le=240)
    servings: int = Field(4, ge=1, le=12)
    meal_type: Optional[MealType] = None
    additional_preferences: Optional[str] = Field(None, max_length=500)

File: zeus-app/src/types/recipe.ts
export interface Ingredient {
  name: string;
  quantity: string;
  unit: string;
  section?: string;
}

export interface Instruction {
  step: number;
  instruction: string;
}

export type DifficultyLevel = 'Easy' | 'Medium' | 'Hard';
export type MealType = 'Breakfast' | 'Lunch' | 'Dinner' | 'Snack' | 'Dessert' | 'Sides';

// Named scenario collection keys understood by the /api/recipes/collections/{key} endpoint.
export type RecipeCollectionKey =
  | 'quick_weeknight'
  | 'quick'
  | 'one_pot'
  | 'sheet_pan'
  | 'slow_cooker'
  | 'instant_pot'
  | 'air_fryer'
  | 'no_cook'
  | 'healthy_light'
  | 'make_ahead'
  | 'meal_prep'
  | 'comfort_food'
  | 'family_favorites';

export interface Recipe {
  id: string;
  user_id: string;
  title: string;
  description?: string;
  image_url?: string;
  ingredients: Ingredient[];
  instructions: Instruction[];
  servings: number;
  prep_time?: number;
  cook_time?: number;
  cuisine_type?: string;
  difficulty: DifficultyLevel;
  meal_type: string[];
  dietary_tags: string[];
  // New uniform tagging dimensions (populated by detection pipeline)
  cooking_method?: string[];   // e.g. ['one_pot'], ['sheet_pan']
  time_tags?: string[];        // e.g. ['quick', 'weeknight']
  style_tags?: string[];       // e.g. ['make_ahead', 'comfort_food']
  is_ai_generated: boolean;
  likes_count: number;
  created_at: string;
  creator_username?: string;
  is_liked?: boolean;
  is_saved?: boolean;
  // Nutrition fields
  calories?: number;
  protein_grams?: number;
  carbs_grams?: number;
  fat_grams?: number;
  serving_size?: string;
}

export interface RecipeCreate {
  title: string;
  description?: string;
  ingredients: Ingredient[];
  instructions: Instruction[];
  servings: number;
  prep_time?: number;
  cook_time?: number;
  cuisine_type?: string;
  difficulty: DifficultyLevel;
  meal_type: MealType[];
  dietary_tags: string[];
  image_url?: string;
}

export interface RecipeFeedFilter {
  cuisine_type?: string;
  cuisine_preferences?: string[];
  difficulty?: DifficultyLevel;
  max_difficulty?: DifficultyLevel;
  max_prep_time?: number;
  meal_type?: MealType;
  dietary_tags?: string[];
  use_pantry_items?: boolean;
  limit?: number;
  offset?: number;
}

